4337/contracts/test/SafeSignerLaunchpad.sol

```diff
@@ -0,0 +1,266 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+pragma solidity >=0.8.0 <0.9.0;
+
+import {IAccount} from "@account-abstraction/contracts/interfaces/IAccount.sol";
+import {UserOperation} from "@account-abstraction/contracts/interfaces/UserOperation.sol";
+import {_packValidationData} from "@account-abstraction/contracts/core/Helpers.sol";
+import {ISignatureValidator} from "@safe-global/safe-contracts/contracts/interfaces/ISignatureValidator.sol";
+import {SafeStorage} from "@safe-global/safe-contracts/contracts/libraries/SafeStorage.sol";
+
+interface IUniqueSignerFactory {
+    /**
+     * @notice Gets the unique signer address for the specified data.
+     * @dev The unique signer address must be unique for some given data. The signer is not guaranteed to be created yet.
+     * @param data The signer specific data.
+     * @return signer The signer address.
+     */
+    function getSigner(bytes memory data) external view returns (address signer);
+
+    /**
+     * @notice Create a new unique signer for the specified data.
+     * @dev The unique signer address must be unique for some given data. This must not revert if the unique owner already exists.
+     * @param data The signer specific data.
+     * @return signer The signer address.
+     */
+    function createSigner(bytes memory data) external returns (address signer);
+
+    /**
+     * @notice Verifies a signature for the specified address without deploying it.
+     * @dev This must be equivalent to first deploying the signer with the factory, and then verifying the signature
+     * with it directly: `factory.createSigner(signerData).isValidSignature(data, signature)`
+     * @param data The data whose signature should be verified.
+     * @param signature The signature bytes.
+     * @param signerData The signer data to verify signature for.
+     * @return magicValue Returns `ISignatureValidator.isValidSignature.selector` when the signature is valid. Reverting or returning any other value implies an invalid signature.
+     */
+    function isValidSignatureForSigner(
+        bytes calldata data,
+        bytes calldata signature,
+        bytes calldata signerData
+    ) external view returns (bytes4 magicValue);
+}
+
+/**
+ * @title SafeOpLaunchpad - A contract for Safe initialization with custom unique signers that would violate ERC-4337 factory rules.
+ * @dev The is intended to be set as a Safe proxy's implementation for ERC-4337 user operation that deploys the account.
+ */
+contract SafeSignerLaunchpad is IAccount, SafeStorage {
+    bytes32 private constant DOMAIN_SEPARATOR_TYPEHASH = keccak256("EIP712Domain(uint256 chainId,address verifyingContract)");
```

Comment:
Didn't you figure out that pre-computed constants are cheaper? 😄

Reply:
Yes, [by around 13 gas units](https://github.com/safe-global/safe-modules/issues/179). Since this is a test contract, I just went for what was easier to code.